python/ray/client_builder.py

```diff
@@ -59,6 +59,21 @@ def connect(self) -> ClientInfo:
             ray_commit=client_info_dict["ray_commit"],
             protocol_version=client_info_dict["protocol_version"])
 
+    def disconnect(self) -> None:
+        """
+        Disconnect Ray. This either disconnects from the remote Client Server
+        or shuts the current driver down.
+        """
+        if ray.util.client.ray.is_connected():
+            # This is only a client connected to a server.
+            ray.util.client_connect.disconnect()
+        elif ray.worker.global_worker.node.is_head():
+            logger.warning("The current Ray Cluster is scoped to this process. "
+            "Disconnecting will shutdown the cluster.")
+        else:
+            # This is only a driver connected to an existing cluster.
+            ray.shutdown()
```

Comment:
@pcmoritz This is the main question I have :)

Reply:
How about this: `ray.client().connect()` returns a `ClientContext` object that has a `.disconnect()` object to disconnect the client. So you can write

```
client = ray.client().connect()
# Do something with client
client.disconnect()
```

In the future we can extend this to
```
with ray.client().connect():
    # Do something with client
```

In fact it might be better to only have the with syntax, and then the `.disconnect()` can be done by calling the `.__exit__()` on the ClientContext. The ClientContext can also have the current .ray_version() etc. information.

The ClientBuilder doesn't have a `.disconnect()` method and also `.disconnect()` never shuts dow the local cluster. The cluster can only be shut down with the `.shutdown()` method of `LocalClientBuilder` (and the AnyscaleClientBuilder or any other ClientBuilder that supports starting clusters).

Reply:
So is your proposal to also add a `.shutdown()` method to the `LocalClientBuilder`? I think that `ray.shutdown()` would just work for that?

Reply:
Yeah and then we can deprecate `ray.shutdown()`. The `.shutdown()` method for the ClientBuilders that support creating clusters is more flexible since it also allows you to shut down a remove cluster that was not created by the same script.

Reply:
I will add `shutdown()` at a later point.